projects/ADuCM3029_demo_cn0428_cn0429/src/ADuCM3029_demo_cn0428_cn0429.cpp

```diff
@@ -0,0 +1,1307 @@
+		exponent = exponent * 2 + (s[i] - '0');
+
+	exponent -= 127;
+
+	if (exponent > -127) {
+		mantissa = 1;
+		exponent -= 23;
+	} else {
+		mantissa = 0;
+		exponent = -126;
+		exponent -= 23;
+	}
+
+	for (i = 9; i <= 31; i++)
+		mantissa = mantissa * 2 + (s[i] - '0');
+
+	f = mantissa;
+
+	while (exponent > 0)
+		f *= 2, exponent--;
+
+	while (exponent < 0)
+		f /= 2, exponent++;
+
+	if (sign)
+		f = -f;
+
+	return f;
+}
+
+/* ================================================== */
+/* 					 State Machine					  */
+/* ================================================== */
+void DataDisplayFSM(void)
+{
+	uint8_t 	readChar = 0;	// variable holding address of selected sensor
+	uint8_t 	cmdResponseBuff[256] = {0};	// buffer holding sensor response to commands
+	uint8_t		tempBuff[40] = {0};			// buffer for EIS results parsing
+	double		EISpartialresult[8] = {0};	// array of partial results (one line) of EIS
+	int32_t		EISpartialDFTresult[8] = {0};	// array of partial results (DFT impedance) of EIS
+	uint32_t	EISresults[192] = {0};		// array of EIS results
+	int16_t 	sigVal = 0;		// temporary variable for signed value
+	uint16_t 	unsigVal = 0;	// temporary variable for unsigned value
+	uint32_t	unsiglVal = 0;	// temporary variable for long unsigned value
+	char		tempChar = 0;   // temporary variable for char
+	bool 		exitloop = false;
+
+	uint8_t pulseDuration, pulseAmplitude;	// Pulse test parameters
+
+	uint8_t tempError;
+	uint8_t numbytes;
+	uint8_t cmdLen;
+	uint16_t timeout;
+
+	switch(FSM_State) {
+	case INIT:
+		FSM_State = COMMAND;
+
+		tempError = SensorInitwithI2CAddr(ADI_CFG_I2C_SENSOR1_ADDR, 1);
+		if (tempError == SENSOR_ERROR_NONE) {
+			S1_detected = true;
+		} else if (tempError == SENSOR_ERROR_PH) {	//Water Quality Board installed
+			S1_detected = true;
+			FSM_State = WATER;
+		}
+
+		tempError = SensorInitwithI2CAddr(ADI_CFG_I2C_SENSOR2_ADDR, 2);
+		if (tempError == SENSOR_ERROR_NONE) {
+			S2_detected = true;
+		} else if (tempError == SENSOR_ERROR_PH) {	//Water Quality Board installed
+			S2_detected = true;
+			FSM_State = WATER;
+		}
+
+		tempError = SensorInitwithI2CAddr(ADI_CFG_I2C_SENSOR3_ADDR, 3);
+		if (tempError == SENSOR_ERROR_NONE) {
+			S3_detected = true;
+		} else if (tempError == SENSOR_ERROR_PH) {	//Water Quality Board installed
+			S3_detected = true;
+			FSM_State = WATER;
+		}
+
+		tempError = SensorInitwithI2CAddr(ADI_CFG_I2C_SENSOR4_ADDR, 4);
+		if (tempError == SENSOR_ERROR_NONE) {
+			S4_detected = true;
+		} else if (tempError == SENSOR_ERROR_PH) {	//Water Quality Board installed
+			S4_detected = true;
+			FSM_State = WATER;
+		}
+
+		if(S1_detected||S2_detected||S3_detected||S4_detected) {
+			UART_TX("----------------------------------------------" _EOS);
+			cmdReceived = 1; //Ensure welcome message printout
+		} else UART_TX("No Sensor Board Found! Please check connections and reset ADICUP3029."
+				       _EOS);
+		break;
+
+	case COMMAND:
+		if (check_string((uint8_t *)"PRESENCE_CHECK", 14)) {
```

Comment:
This check_string here is a bit redundant. You already parse string in CLI i don't quite see the need for the ring buffer and another function to parse them another time.  I would consider something similar to this where you parse the commands and do the processing there. https://github.com/analogdevicesinc/EVAL-ADICUP360/blob/master/projects/ADuCM360_demo_cn0411/src/CN0411.c

Reply:
There is some history here. We added the CLI late in this project when we decided we weren't releasing a GUI. The CN0428 was already built assuming the user could interract with the command line, but the CN0429 GUI passed command strings and arguments that aren't particularly easy for a user to type (e.g., RUN_PULSE expects direct hex values for amplitude and duration as arguments, many of which are not available on a keyboard) and with the assumption that the GUI wouldn't send incorrect commands or arguments. We added the CLI layer in order to simplify the user input, add some error checking, and translate that into the existing command structure. I think Michal chose not to make major changes to the state machine in order to reduce effort and risk at that stage of the project.